tests/plot_sync/test_plot_sync.py

```diff
@@ -0,0 +1,540 @@
+    return (
+        sender._last_sync_id != previous_last_sync_id
+        and sender._last_sync_id == receiver._last_sync_id != 0
+        and receiver.state() == State.idle
+        and not sender._lock.locked()
+    )
+
+
+def assert_path_list_matches(expected_list: List, actual_list: List):
+    assert len(expected_list) == len(actual_list)
+    for item in expected_list:
+        assert str(item) in actual_list
+
+
+@dataclass
+class ExpectedResult:
+    valid_count: int = 0
+    valid_delta: PlotListDelta = field(default_factory=PlotListDelta)
+    invalid_count: int = 0
+    invalid_delta: PathListDelta = field(default_factory=PathListDelta)
+    keys_missing_count: int = 0
+    keys_missing_delta: PathListDelta = field(default_factory=PathListDelta)
+    duplicates_count: int = 0
+    duplicates_delta: PathListDelta = field(default_factory=PathListDelta)
+    callback_passed: bool = False
+
+    def reset(self) -> None:
+        self.valid_count = 0
+        self.valid_delta.clear()
+        self.invalid_count = 0
+        self.invalid_delta.clear()
+        self.keys_missing_count = 0
+        self.keys_missing_delta.clear()
+        self.duplicates_count = 0
+        self.duplicates_delta.clear()
+        self.callback_passed = False
+
+    def add_valid(self, list_plots: List) -> None:
+        self.valid_count += len(list_plots)
+        self.valid_delta.additions.update({x.prover.get_filename(): x for x in list_plots})
+
+    def remove_valid(self, list_paths: List) -> None:
+        self.valid_count -= len(list_paths)
+        self.valid_delta.removals += list_paths.copy()
+
+    def add_invalid(self, list_paths: List) -> None:
+        self.invalid_count += len(list_paths)
+        self.invalid_delta.additions += list_paths.copy()
+
+    def remove_invalid(self, list_paths: List) -> None:
+        self.invalid_count -= len(list_paths)
+        self.invalid_delta.removals += list_paths.copy()
+
+    def add_keys_missing(self, list_paths: List) -> None:
+        self.keys_missing_count += len(list_paths)
+        self.keys_missing_delta.additions += list_paths.copy()
+
+    def remove_keys_missing(self, list_paths: List) -> None:
+        self.keys_missing_count -= len(list_paths)
+        self.keys_missing_delta.removals += list_paths.copy()
+
+    def add_duplicates(self, list_paths: List) -> None:
+        self.duplicates_count += len(list_paths)
+        self.duplicates_delta.additions += list_paths.copy()
+
+    def remove_duplicates(self, list_paths: List) -> None:
+        self.duplicates_count -= len(list_paths)
+        self.duplicates_delta.removals += list_paths.copy()
+
+
+@dataclass
+class Environment:
+    root_path: Path
+    harvester_services: List[Service]
+    farmer_service: Service
+    harvesters: List[Harvester]
+    farmer: Farmer
+    dir_1: TestDirectory
+    dir_2: TestDirectory
+    dir_3: TestDirectory
+    dir_4: TestDirectory
```

Comment:
Whats the advantage here over having named entries? Imo `env.dir_invalid` looks better than `env.directories[4]`. Of course we could add an enum with the name mapping but then i wonder again why it would be better :)